### DISCUSSION TITLE
Higher memory (RAM) support

### DISCUSSION THREAD

**Discussion by darvin** [dbalsom/marty#20] [Ideas]
How hard would it be to extend emulator to support 640k+ memory? Which mode would be the easiest to code correctly target?

**Comment by dbalsom** [comment]
Two options I am considering are implementing the Lo-tech EMS board https://www.lo-tech.co.uk/wiki/Lo-tech_2MB_EMS_Board
or something more period-accurate like the Intel Above Board, if I can find documentation

**Reply by darvin** [reply]
do you think its necessity? is it possible to circumvent it and just increate the amount of the exposed memory, allowing operating system to handle it however it wishes? Obviously, total majority of applications will fail to notice the exposed memory: 640k barrier is hardcoded in DOS. However those applications that are informed about presence of additional memory over first 640k will be able to access it.

is there a.reasoning behind true to metal implementation of the memory?

**Reply by dbalsom** [reply]
If you use GLaBIOS you can use 736K of memory with MartyPC, but you bump up against the memory map of the CGA card, and above that there could be other devices mapped and option ROMs, so it is not practical to treat the address space there as RAM. 

Typically, to access more than conventional memory in DOS, there were two standards, EMS and XMS, which for an 8088 machine required appropriate hardware and drivers.

**Reply by google0101-ryan** [reply]
Because no program can possibly use this extra RAM. This is before the times of the a20 line, so programs had a hardcoded limit of `0xFFFFF`, and addresses higher than that wrap around to 0. So it is not possible to "just increase the amount of exposed memory." However, this is remedied with the 80286 and newer.

At least, that's how I see it

**Comment by dbalsom** [comment]
I've finally added support for an EMS card in 
https://github.com/dbalsom/martypc/commit/5948f12b08a1e6f088d57e5d38bcb528e14265f5

Now you can add 2MB of EMS memory to your PC/XT
